### PR TITLE
Complete unit test coverage for upgradeable renouncable proxy

### LINKED ISSUE #90: write additional unit tests for proxy

### PR DESCRIPTION
**Pull request by roleengineer**
Two mock contracts were implemented to test the behavior of the upgradeable renounceable proxy under different scenarios:

- Mock with Selector Clashes: This mock simulates cases where there are selector clashes between the proxy and implementation, allowing us to test how the proxy handles such situations.
- Mock with Stateful Implementation: This mock simulates an implementation which works with state, allowing us to evaluate how the proxy handles state.

Tested:

- Proxy's Native Functions: all native proxy functions to ensure their correctness in various scenarios, including when selector clashes or a stateful implementation are present.
- Behavior with Clashing Implementations: how the proxy behaves when the implementation contract has selector clashes, leading to potential vulnerabilities or unexpected behavior.
- Behavior with Stateful Implementations: how the proxy behaves with a stateful implementation, ensuring it handles state correctly.

During testing, a serious issue was discovered where the proxy's native functionality can be bypassed by a malicious implementation. Specifically: 

Proxy Admin & Implementation: any address (not just the proxy admin) can change the proxy's admin and implementation. This is a critical vulnerability, if we have a constraint that only proxy native functionality can change admin and implementation.

Given the findings, conclusion is that fixes are required in the proxy contract to address these security issues and ensure the integrity of the upgradeable renouncable proxy functionality.

### CODE REVIEW

**Comment by roleengineer** [comment]
solves https://github.com/aboutcircles/circles-contracts-v2/issues/90

File: test/groups/upgradeableProxy/adminOperationsUpgradeableRenounceableProxy.t.sol [hunk 1]
```diff
@@ -1,23 +1,37 @@
...
-import "../../../src/groups/UpgradeableRenounceableProxy.sol";
-import "../../../src/errors/Errors.sol";
-import "../groupSetup.sol";
+import {console2, Test} from "forge-std/Test.sol";
+import "src/errors/Errors.sol";
+import "test/groups/groupSetup.sol";
+import {ERC1967Utils} from "@openzeppelin/contracts/proxy/ERC1967/ERC1967Utils.sol";
+import {
+    UpgradeableRenounceableProxy, IUpgradeableRenounceableProxy
+} from "src/groups/UpgradeableRenounceableProxy.sol";
```
Comment (benjaminbollen) [inline test/groups/upgradeableProxy/adminOperationsUpgradeableRenounceableProxy.t.sol:10]: I can't believe I never realised using an absolute path is so much more readable and better. Internal style-guide updated !

File: test/groups/upgradeableProxy/adminOperationsUpgradeableRenounceableProxy.t.sol [hunk 2]
```diff
@@ -120,4 +346,42 @@ contract adminOperationsUpgradeableRenounceableProxy is Test, GroupSetup {
         uint256 balanceAfter = hub.balanceOf(_minter, tokenIdGroup);
         assertEq(balanceAfter, balanceBefore + _amount);
     }
+
+    /// @dev makes admin (set to group) upgradeToAndCall call until admin is not renounced
```
Comment (benjaminbollen) [inline test/groups/upgradeableProxy/adminOperationsUpgradeableRenounceableProxy.t.sol]: typo? "...until admin is renounced"?

File: test/groups/upgradeableProxy/adminOperationsUpgradeableRenounceableProxy.t.sol [hunk 3]
```diff
@@ -79,24 +127,202 @@ contract adminOperationsUpgradeableRenounceableProxy is Test, GroupSetup {
...
+        assertEq(keccak256(returnedBytes), keccak256("newMintPolicy"));
+        vm.stopPrank();
+    }
+
+    // External receive()
+
+    function testReceive(address anyAddress) public {
+        vm.deal(anyAddress, 1 ether);
+        vm.expectRevert(UpgradeableRenounceableProxy.BlockReceive.selector);
+        (bool success,) = address(proxy).call{value: 1 ether}("");
```
Comment (benjaminbollen) [inline test/groups/upgradeableProxy/adminOperationsUpgradeableRenounceableProxy.t.sol:280]: shouldn't `anyAddress` be sending value 1 eth?

**Review by benjaminbollen — COMMENTED** [review]
Great work!

I left small questions, you can see whether they need resolving or not; happy to merge following your response

File: test/groups/upgradeableProxy/adminOperationsUpgradeableRenounceableProxy.t.sol (hunk 2 above)
Comment (roleengineer) [reply test/groups/upgradeableProxy/adminOperationsUpgradeableRenounceableProxy.t.sol]: sorry, formulated badly
```suggestion
    /// @dev calls the upgradeToAndCall function by the proxy admin (works until admin is renounced)
```

File: test/groups/upgradeableProxy/adminOperationsUpgradeableRenounceableProxy.t.sol (hunk 3 above)
Comment (roleengineer) [reply test/groups/upgradeableProxy/adminOperationsUpgradeableRenounceableProxy.t.sol:280]: good catch! missed one line!
```suggestion
        vm.prank(anyAddress);
        (bool success,) = address(proxy).call{value: 1 ether}("");
```

**Comment by benjaminbollen** [comment]
both questions resolved. LGTM!

**Review by benjaminbollen — APPROVED** [review]

**Pull request merged**